File: javascript/src/method/forEach.ts
//··························································
// Copyright (c) 2023-2025. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {CollectionHolder}           from "../CollectionHolder"
import type {MinimalistCollectionHolder} from "../MinimalistCollectionHolder"
import type {ValueIndexCallback}         from "../type/callback"

import {isArray}                       from "./isArray"
import {isArrayByStructure}            from "./isArrayByStructure"
import {isCollectionHolder}            from "./isCollectionHolder"
import {isCollectionHolderByStructure} from "./isCollectionHolderByStructure"

//#region -------------------- Facade method --------------------

/**
 * Perform a given {@link action} on each element
 *
 * @param collection The {@link Nullable nullable} collection ({@link MinimalistCollectionHolder}, {@link CollectionHolder} or {@link ReadonlyArray Array})
 * @param action     The given action
 * @see ReadonlyArray.forEach
 * @see ReadonlySet.forEach
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html Kotlin forEach(action)
 * @see https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/Iterable.html#forEach(java.util.function.Consumer) Java forEach(action)
 * @extensionFunction
 */
export function forEach<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | readonly T[]>, action: ValueIndexCallback<T>,): void {
    if (collection == null)
        return
    if (isCollectionHolder<T>(collection,))
        forEachByCollectionHolder(collection, action,)
    else if (isArray(collection,))
        forEachByArray(collection, action,)
    else if (isCollectionHolderByStructure<T>(collection,))
        forEachByCollectionHolder(collection, action,)
    else if (isArrayByStructure<T>(collection,))
        forEachByArray(collection, action,)
    else
        forEachByMinimalistCollectionHolder(collection, action,)
}


/**
 * Perform a given {@link action} on each element
 *
 * @param collection The {@link MinimalistCollectionHolder collection}
 * @param action     The given action
 * @see ReadonlyArray.forEach
 * @see ReadonlySet.forEach
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html Kotlin forEach(action)
 * @see https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/Iterable.html#forEach(java.util.function.Consumer) Java forEach(action)
 * @extensionFunction
 */
export function forEachByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, action: ValueIndexCallback<T>,): void {
    if (collection == null)
        return

    const size = collection.size
    if (size == 0)
        return
    if (action.length == 1)
        __with1Argument(collection, action as (value: T,) => void, size,)
    else if (action.length >= 2)
        __with2Argument(collection, action, size,)
    else
        __with0Argument(action as () => void, size,)
}

/**
 * Perform a given {@link action} on each element
 *
 * @param collection The {@link CollectionHolder collection}
 * @param action     The given action
 * @see ReadonlyArray.forEach
 * @see ReadonlySet.forEach
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html Kotlin forEach(action)
 * @see https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/Iterable.html#forEach(java.util.function.Consumer) Java forEach(action)
 * @extensionFunction
 */
export function forEachByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, action: ValueIndexCallback<T>,): void {
    if (collection == null)
        return
    if (collection.isEmpty)
        return
    if (action.length == 1)
        __with1Argument(collection, action as (value: T,) => void, collection.size,)
    else if (action.length >= 2)
        __with2Argument(collection, action, collection.size,)
    else
        __with0Argument(action as () => void, collection.size,)
}

/**
 * Perform a given {@link action} on each element
 *
 * @param collection The {@link ReadonlyArray collection}
 * @param action     The given action
 * @see ReadonlyArray.forEach
 * @see ReadonlySet.forEach
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html Kotlin forEach(action)
 * @see https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/Iterable.html#forEach(java.util.function.Consumer) Java forEach(action)
 * @extensionFunction
 */
export function forEachByArray<const T, >(collection: Nullable<readonly T[]>, action: ValueIndexCallback<T>,): void {
    if (collection == null)
        return

    const size = collection.length
    if (size == 0)
        return
    if (action.length == 1)
        __with1ArgumentByArray(collection, action as (value: T,) => void, size,)
    else if (action.length >= 2)
        __with2ArgumentByArray(collection, action, size,)
    else
        __with0Argument(action as () => void, size,)
}

//#endregion -------------------- Facade method --------------------
//#region -------------------- Loop methods --------------------

function __with0Argument(action: () => void, size: number,) {
    let index = size
    while (index-- > 0)
        action()
}


function __with1Argument<const T, >(collection: MinimalistCollectionHolder<T>, action: (value: T,) => void, size: number,) {
    let index = -1
    while (++index < size)
        action(collection.get(index,),)
}

function __with1ArgumentByArray<const T, >(collection: readonly T[], action: (value: T,) => void, size: number,) {
    let index = -1
    while (++index < size)
        action(collection[index] as T,)
}


function __with2Argument<const T, >(collection: MinimalistCollectionHolder<T>, action: (value: T, index: number,) => void, size: number,) {
    let index = -1
    while (++index < size)
        action(collection.get(index,), index,)
}

function __with2ArgumentByArray<const T, >(collection: readonly T[], action: (value: T, index: number,) => void, size: number,) {
    let index = -1
    while (++index < size)
        action(collection[index] as T, index,)
}

//#endregion -------------------- Loop methods --------------------

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/ForEach.java
package joookiwi.collection.java.method;

import java.util.function.Consumer;
import java.util.function.ObjIntConsumer;
import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;

@NotNullByDefault
public final class ForEach
        extends Utility {

    @Contract(ALWAYS_FAIL_0)
    private ForEach() { throw new ImpossibleConstructionException("The utility class “ForEach” cannot be constructed.", ForEach.class); }

    //#region -------------------- Facade methods --------------------

    //#region -------------------- predicate (T, int) → boolean --------------------

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">JavaScript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                            final ObjIntConsumer<? super T> action) {
        if (collection == null)
            return;

        final var size = collection.size();
        if (size == 0)
            return;
        __with2Argument(collection, action, size);
    }

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">JavaScript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final @Nullable CollectionHolder<? extends T> collection,
                                                            final ObjIntConsumer<? super T> action) {
        if (collection == null)
            return;
        if (collection.isEmpty())
            return;
        __with2Argument(collection, action, collection.size());
    }

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">JavaScript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final T @Nullable @Unmodifiable [] collection,
                                                            final ObjIntConsumer<? super T> action) {
        if (collection == null)
            return;

        final var size = collection.length;
        if (size == 0)
            return;
        __with2Argument(collection, action, size);
    }

    //#endregion -------------------- predicate (T, int) → boolean --------------------
    //#region -------------------- predicate (T) → boolean --------------------

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">JavaScript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                            final Consumer<? super T> action) {
        if (collection == null)
            return;

        final var size = collection.size();
        if (size == 0)
            return;
        __with1Argument(collection, action, size);
    }

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">JavaScript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final @Nullable CollectionHolder<? extends T> collection,
                                                            final Consumer<? super T> action) {
        if (collection == null)
            return;
        if (collection.isEmpty())
            return;
        __with1Argument(collection, action, collection.size());
    }

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">JavaScript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final T @Nullable @Unmodifiable [] collection,
                                                            final Consumer<? super T> action) {
        if (collection == null)
            return;

        final var size = collection.length;
        if (size == 0)
            return;
        __with1Argument(collection, action, size);
    }

    //#endregion -------------------- predicate (T) → boolean --------------------
    //#region -------------------- predicate () → boolean --------------------

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">JavaScript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">Javascript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                            final Runnable action) {
        if (collection == null)
            return;

        final var size = collection.size();
        if (size == 0)
            return;
        __with0Argument(action, size);
    }

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">Javascript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">Javascript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final @Nullable CollectionHolder<? extends T> collection,
                                                            final Runnable action) {
        if (collection == null)
            return;
        if (collection.isEmpty())
            return;
        __with0Argument(action, collection.size());
    }

    /// Perform a given `action` on each element
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param action     The given action
    /// @see Iterable#forEach(Consumer)
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach">Javascript ReadonlyArray.forEach</a>
    /// @see <a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Set/forEach">Javascript ReadonlySet.forEach</a>
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/for-each.html">Kotlin forEach(action)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> void forEach(final T @Nullable @Unmodifiable [] collection,
                                                            final Runnable action) {
        if (collection == null)
            return;

        final var size = collection.length;
        if (size == 0)
            return;
        __with0Argument(action, size);
    }

    //#endregion -------------------- predicate () → boolean --------------------

    //#endregion -------------------- Facade methods --------------------
    //#region -------------------- Loop methods --------------------

    private static void __with0Argument(final Runnable action,
                                        final int size) {
        var index = size;
        while (index-- > 0)
            action.run();
    }


    private static <T extends @Nullable Object> void __with1Argument(final MinimalistCollectionHolder<? extends T> collection,
                                                                     final Consumer<? super T> action,
                                                                     final int size) {
        var index = -1;
        while (++index < size)
            action.accept(collection.get(index));
    }

    private static <T extends @Nullable Object> void __with1Argument(final T [] collection,
                                                                     final Consumer<? super T> action,
                                                                     final int size) {
        var index = -1;
        while (++index < size)
            action.accept(collection[index]);
    }


    private static <T extends @Nullable Object> void __with2Argument(final MinimalistCollectionHolder<? extends T> collection,
                                                                     final ObjIntConsumer<? super T> action,
                                                                     final int size) {
        var index = -1;
        while (++index < size)
            action.accept(collection.get(index), index);
    }

    private static <T extends @Nullable Object> void __with2Argument(final T [] collection,
                                                                     final ObjIntConsumer<? super T> action,
                                                                     final int size) {
        var index = -1;
        while (++index < size)
            action.accept(collection[index], index);
    }

    //#endregion -------------------- Loop methods --------------------

}
